src/server/robot/robot-manager.ts

```diff
@@ -0,0 +1,57 @@
+import { GridIndices } from "./grid-indices";
+import { Robot } from "./robot";
+
+/**
+ * Stores robots. Provides utilities for finding them by position.
+ */
+export class RobotManager {
+    /**
+     * Maps robot ids to robots.
+     */
+    idsToRobots: Map<string, Robot> = new Map();
+
+    /**
+     * Maps robot locations to their ids.
+     */
+    indicesToIds: Map<GridIndices, string> = new Map();
+
+    constructor(robots: Robot[]) {
+        robots.forEach((robot) => this.addRobot(robot));
+    }
+
+    addRobot(robot: Robot) {
+        this.idsToRobots.set(robot.id, robot);
+        this.indicesToIds.set(robot.homeIndices, robot.id);
+    }
+
+    /**
+     * Retrieves a robot by id.
+     * Throws if no robot is found.
+     */
+    getRobot(robotId: string): Robot {
+        const robot = this.idsToRobots.get(robotId);
+        if (robot === undefined) {
+            throw new Error("Failed to find robot with id " + robotId);
```

Comment:
Honestly not a big fan of throwing errors because it crashes the server if uncaught (try-catches are honestly extremely annoying because it creates a new scope inside the try block)

Reply:
I guess my general stance on errors is it's best to throw one unless one of the common scenarios for the API is going in with a plan to handle errors gratefully. In our case, I think most of the time we expect not finding a valid robot to crash since that otherwise means e.g. a command is getting sent into the ether which isn't something I'm too worried about catching. `try-catch` is nice because it allows errors to propagate up scopes without manual help, which prevents code from getting invaded with constant `undefined` checks and `| undefined` return types (annoying, verbose) or `!` assertions (sketchy, arguably worse than just throwing an error). 

I could also see an argument for a safe and unsafe version of these types of methods to give the caller the manual option to choose. I'm also open to deferring this until we know more about the usage. I suspect a lot of disconnection errors are best handled at the top level (e.g. the API handler) - I'm not sure of more specific methods we could take to recover from, e.g. a bad robotId inside a command.